engine/src/flutter/impeller/renderer/pipeline_library.h

```diff
@@ -74,13 +75,31 @@ class PipelineLibrary : public std::enable_shared_from_this<PipelineLibrary> {
   virtual void RemovePipelinesWithEntryPoint(
       std::shared_ptr<const ShaderFunction> function) = 0;
 
+  void LogPipelineUsage(const PipelineDescriptor& p);
+
+  void LogPipelineCreation(const PipelineDescriptor& p);
+
+  std::unordered_map<PipelineDescriptor,
+                     int,
+                     ComparableHash<PipelineDescriptor>,
+                     ComparableEqual<PipelineDescriptor>>
+  GetPipelineUseCounts();
+
  protected:
   PipelineLibrary();
 
  private:
   PipelineLibrary(const PipelineLibrary&) = delete;
 
   PipelineLibrary& operator=(const PipelineLibrary&) = delete;
+
+  std::unordered_map<PipelineDescriptor,
```

Comment:
Please use the synchronization utilities from `//impeller/base/thread.h`. These, along with the static thread safety annotation on the resources the mutexes will guard give you static thread safety checks.

An example that I just looked up is here https://github.com/flutter/flutter/blob/f63cf73d310065a11b67d82ba88618e342227c7c/engine/src/flutter/impeller/playground/backend/gles/playground_impl_gles.cc#L49. Notice the `RWMutex` along with the `IPLR_GUARDED_BY` on the resource the mutex guards.